File: SQL/005. Longest Palindromic Substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        string str = "", ans = "";
        int len = s.length();
        int maxl = -1, cnt;
        for (int i = 0; i < len; i++) {
            str += '#';
            str += s[i];
        }
        str += '#';
        for (int i = 1; i < 2 * len; i++) {
            cnt = 0;
            while ((i - cnt >= 0) && (i + cnt <= 2 * len) && (str[i - cnt] == str[i + cnt]))
                cnt++;
            cnt--;
            if (cnt > maxl) {
                maxl = cnt;
                ans = s.substr((i - cnt) / 2, (i + cnt) / 2 - (i - cnt) / 2);
            }
        }
        return ans;
        
    }
};

File: python/005. Longest Palindromic Substring.py
class Solution:
    # @return a string
    def getlongestpalindrome(self, s, l, r):
        while l >= 0 and r < len(s) and s[l] == s[r]:
            l -= 1; r += 1
        return l+1, r-1
    def longestPalindrome(self, s):
        start = end = 0
        for i in range(len(s)):
            l, r = self.getlongestpalindrome(s, i, i)
            if r - l > end - start:
                start = l
                end = r
            l, r = self.getlongestpalindrome(s, i, i + 1)
            if r - l > end - start: 
                start = l
                end = r
        return s[start:end+1]

